library_generation/generate_pr_description.py

```diff
@@ -0,0 +1,199 @@
+    file paths will be considered.
+    Note that baseline_commit should be an ancestor of latest_commit.
+
+    :param repo_url: the url of the repository.
+    :param latest_commit: the newest commit to be considered in
+    selecting commit message.
+    :param baseline_commit: the oldest commit to be considered in
+    selecting commit message. This commit should be an ancestor of
+    :param paths: a mapping from file paths to library_name.
+    :param generator_version: the version of the generator.
+    :param is_monorepo: whether to generate commit messages in a monorepo.
+    :return: commit messages.
+    """
+    tmp_dir = "/tmp/repo"
+    shutil.rmtree(tmp_dir, ignore_errors=True)
+    os.mkdir(tmp_dir)
+    repo = Repo.clone_from(repo_url, tmp_dir)
+    commit = repo.commit(latest_commit)
+    qualified_commits = {}
+    while str(commit.hexsha) != baseline_commit:
+        commit_and_name = __filter_qualified_commit(paths=paths, commit=commit)
+        if commit_and_name != ():
+            qualified_commits[commit_and_name[0]] = commit_and_name[1]
+        commit_parents = commit.parents
+        if len(commit_parents) == 0:
+            break
+        commit = commit_parents[0]
+    shutil.rmtree(tmp_dir, ignore_errors=True)
+    return __combine_commit_messages(
+        latest_commit=latest_commit,
+        baseline_commit=baseline_commit,
+        commits=qualified_commits,
+        generator_version=generator_version,
+        is_monorepo=is_monorepo,
+    )
+
+
+def __filter_qualified_commit(paths: Dict[str, str], commit: Commit) -> (Commit, str):
+    """
+    Returns a tuple of a commit and libray_name.
+    A qualified commit means at least one file changes in that commit is
+    within the versioned proto_path in paths.
+
+    :param paths: a mapping from versioned proto_path to library_name.
+    :param commit: a commit under consideration.
+    :return: a tuple of a commit and library_name if the commit is
+    qualified; otherwise an empty tuple.
+    """
+    for file in commit.stats.files.keys():
+        versioned_proto_path = find_versioned_proto_path(file)
+        if versioned_proto_path in paths:
+            return commit, paths[versioned_proto_path]
+    return ()
+
+
+def __combine_commit_messages(
+    latest_commit: str,
+    baseline_commit: str,
+    commits: Dict[Commit, str],
+    generator_version: str,
+    is_monorepo: bool,
+) -> str:
+    messages = [
+        f"This pull request is generated with proto changes between googleapis commit {baseline_commit} (exclusive) and {latest_commit} (inclusive).",
+        "Qualified commits are:",
+    ]
+    for commit in commits:
+        short_sha = commit.hexsha[:7]
+        messages.append(
+            f"[googleapis/googleapis@{short_sha}](https://github.com/googleapis/googleapis/commit/{commit.hexsha})"
+        )
+
+    for commit, library_name in commits.items():
+        convention, _, summary = commit.message.partition(":")
```

Comment:
We need to consider a scenario that the commit message may include multiple conventional commits.

Reply:
Done, verified in unit test.